decred/decred/dcr/addrlib.py

```diff
@@ -0,0 +1,447 @@
+        a txout's script.
+
+        Returns:
+            ByteArray: The script address.
+        """
+        return self.serialize()
+
+    def hash160(self):
+        """
+        The hash160 of the serialized pubkey.
+
+        Returns:
+            ByteArray: The hash.
+        """
+        return hash160(self.serialize().bytes())
+
+
+class AddressScriptHash(Address):
+    """
+    AddressScriptHash is an Address for a pay-to-script-hash (P2SH) transaction.
+    """
+
+    def __init__(self, scriptHash, netParams):
+
+        if len(scriptHash) != RIPEMD160_SIZE:
+            raise DecredError(f"incorrect script hash length {len(scriptHash)}")
+
+        super().__init__(netParams)
+        self.netID = netParams.ScriptHashAddrID
+        self.scriptHash = scriptHash
+
+    @staticmethod
+    def fromScript(script, netParams):
+        """
+        Create a new AddressScriptHash from a redeem script.
+
+        Args:
+            script (ByteArray): the redeem script
+            netParams (module): the network parameters
+
+        Returns:
+            AddressScriptHash: An address object.
+        """
+        return AddressScriptHash(hash160(script.b), netParams)
+
+    def __eq__(self, a):
+        """Check that other address is equivalent to this address."""
+        if isinstance(a, str):
+            return a == self.string()
+        elif isinstance(a, AddressScriptHash):
+            return a.scriptHash == self.scriptHash and a.netID == self.netID
+        return False
+
+    def string(self):
+        """
+        A base-58 encoding of the pubkey hash.
+
+        Returns:
+            str: The encoded address.
+        """
+        return encodeAddress(self.netID, self.scriptHash)
+
+    def address(self):
+        """
+        The string encoding of a pay-to-script-hash address.
+
+        Returns:
+            str: The encoded address.
+        """
+        return self.string()
+
+    def scriptAddress(self):
+        """
+        The raw bytes of the address to be used when inserting the address into
+        a txout's script.
+
+        Returns:
+            ByteArray: The script address.
+        """
+        return self.scriptHash.copy()
+
+    def hash160(self):
+        """
+        A copy of the scriptHash.
+
+        Returns:
+            ByteArray: The hash.
+        """
+        return self.scriptHash.copy()
+
+
+class AddressEdwardsPubKey(Address):
+    """unimplemented"""
+
+    def __init__(*a, **k):
+        raise NotImplementedError("AddressEdwardsPubKey not implemented")
+
+
+class AddressSecSchnorrPubKey(Address):
+    """unimplemented"""
+
+    def __init__(*a, **k):
+        raise NotImplementedError("AddressSecSchnorrPubKey implemented")
+
+
+def encodeAddress(netID, k):
+    """
+    Base-58 encode the number, with the netID prepended byte-wise.
+
+    Args:
+        netID (byte-like): The addresses network encoding ID.
+        k (ByteArray): The pubkey or pubkey-hash or script-hash.
```

Comment:
Maybe the two arguments could be swapped. Net data is usually last, and that seems to be stronger than the fact that the netID is prepended in the output.